README.md
# @0x5e/homebridge-tuya-platform

[![npm](https://badgen.net/npm/v/@0x5e/homebridge-tuya-platform)](https://npmjs.com/package/@0x5e/homebridge-tuya-platform)
[![npm](https://badgen.net/npm/dt/@0x5e/homebridge-tuya-platform)](https://npmjs.com/package/@0x5e/homebridge-tuya-platform)
[![mit-license](https://badgen.net/npm/license/@0x5e/homebridge-tuya-platform)](https://github.com/0x5e/homebridge-tuya-platform/blob/main/LICENSE)
[![verified-by-homebridge](https://badgen.net/badge/homebridge/verified/purple)](https://github.com/homebridge/homebridge/wiki/Verified-Plugins)

Tuya plugin for Hombridge, maintained by @0x5e, former employee of Tuya.

Published as [@0x5e/homebridge-tuya-platform](https://npmjs.com/package/@0x5e/homebridge-tuya-platform), currently in beta version.

If beta version works fine for a while, it will be merged into the upstream repo in the future.

## Features

- Optimized code, improved code readability and maintainability.
- Improved stability.
- Less duplicate code.
- Less API errors.
- Less development costs for new accessory categroies.
- More supported devices.
    - Air Quality Sensor
    - Carbon Monoxide Sensor
    - Carbon Dioxide Sensor
    - Motion Sensor
    - Light Sensor
    - Water Detector
    - Temperature and Humidity Sensor
    - Window
- Reimplemented accessory code. Some bug fixed.
    - Switch
    - Outlet
    - Lightbulb
    - Garage Door Opener
    - Window Covering
    - Smoke Sensor
    - Contact Sensor
    - Leak Sensor
- `username` and `password` options are no longer need. However for `Smart Home` project, now `endpoint` and `uid` is required.


## Supported Tuya Devices
See [SUPPORTED_DEVICES.md](./SUPPORTED_DEVICES.md)


## Changelogs
See [CHANGELOG.md](./CHANGELOG.md)


## Installation
Before use, please uninstall `homebridge-tuya-platform` first. They can't run together. (But the config is compatible, no need to delete.)

#### For Homebridge Web UI Users
Go to plugin page, search `@0x5e/homebridge-tuya-platform` and install.


#### For Homebridge Command Line Users

```
npm install @0x5e/homebridge-tuya-platform
```


## Configuration

There's two type of project: `Custom` and `Smart Home`.
The differenct between them is:
- `Custom` Project pull devices from project's asset.
- `Smart Home` Project pull devices from Tuya App user's home.

If you are personal user and don't know which to choose, please use `Smart Home`.

Before configuration, please goto [Tuya IoT Platform](https://iot.tuya.com)
- Create a cloud develop project.
- Go to `Project Page` > `Devices Panel` > `Link Tuya App Account`, link your app account.

## For "Custom" Project

- `platform` - **required** : Must be 'TuyaPlatform'
- `options.projectType` - **required** : Must be '1'
- `options.endpoint` - **required** : Endpoint URL from [API Reference -> Endpoints](https://developer.tuya.com/en/docs/iot/api-request?id=Ka4a8uuo1j4t4#title-1-Endpoints) table.
- `options.accessId` - **required** : Access ID from [Tuya IoT Platform -> Cloud Develop](https://iot.tuya.com/cloud)
- `options.accessKey` - **required** : Access Secret from [Tuya IoT Platform -> Cloud Develop](https://iot.tuya.com/cloud)

## For "Smart Home" Project

- `platform` - **required** : Must be 'TuyaPlatform'
- `options.projectType` - **required** : Must be '2'
- `options.endpoint` - **required** : Endpoint URL from [API Reference -> Endpoints](https://developer.tuya.com/en/docs/iot/api-request?id=Ka4a8uuo1j4t4#title-1-Endpoints) table.
- `options.accessId` - **required** : Access ID from [Tuya IoT Platform -> Cloud Develop](https://iot.tuya.com/cloud)
- `options.accessKey` - **required** : Access Secret from [Tuya IoT Platform -> Cloud Develop](https://iot.tuya.com/cloud)
- `options.uid` - **required** : User ID of the linked Tuya App user.


## Troubleshooting

#### Enable Debug Mode

For Homebridge Web UI users:
- Go to the `Homebridge Setting` page
- Turn on the `Homebridge Debug Mode -D` switch
- Restart HomeBridge.

For Homebridge Command Line Users:
- Start Homebridge with `-D` flag: `homebridge -D`


## Contributing

Please see https://github.com/homebridge/homebridge-plugin-template

PRs and issues are welcome.

# 
Thank you for spend time using the project. If it helps you, don't hesitate to give it a star 🌟:-)


CHANGELOG.md
# Changelog

## [1.7.0] - (unreleased)

### Added
- Add scene support. (#118)
- Add Wireless Switch support (`wxkg`).
- Add Solar Light support (`tyndj`).
- Add Dehumidifier support (`cs`).
- Add Scene Switch support (`wxkg`).
- Add device overriding config support. "Non-standard DP" devices have possibility to be supported now. (#119)
- Add Camera support (`sp`). Thanks @ErrorErrorError for the contribution
- Add Air Conditioner support (`kt`). (#160)
- Add Air Conditioner Controller support (`ktkzq`). (#160)


### Fixed
- Fix `RotationSpeed` missing one level. (#170)
- Fix `bright_value` not sent for the `C/CW` lights who doesn't have `work_mode`. (#171)


### Changed
- Support Ceiling Fan icon customize and Floor Fan `lock`, `swing` feature. (#131)
- Adjust humidity range of dehumidifier and humidifier.
- Print scene id in logs.
- Update support for RGB Power Switch (`dj`).
- Support `countdown` for valve.


## [1.6.0] - (2022.12.3)

This version has been completely rewritten in TypeScript, brings a lot of bug fix and new device support.

### New Accessories
- Add CO Detector support (`cobj`).
- Add CO2 Detector support (`co2bj`).
- Add Water Detector support (`sj`).
- Add Temperature and Humidity Sensor support (`wsdcg`, `wnykq`). Thanks @bimusiek for the contribution
- Add Light Sensor support (`ldcg`).
- Add Motion Sensor support (`pir`).
- Add PM2.5 Detector support (`pm25`).
- Add Door and Window Controller support (`mc`).
- Add Curtain Switch support (`clkg`). (#8)
- Add Human Presence Sensor support (`hps`). (#17)
- Add Thermostat support (`wk`). (#19) Thanks @burcadoruciprian for the contribution
- Add Spotlight support (`sxd`). (#21)
- Add Irrigator support (`ggq`). (#28)
- Add Scene Light Socket support (`qjdcz`). (#33)
- Add Ceiling Fan Light support (`fsd`). (#37)
- Add Thermostat Valve support (`wkf`). (#50)
- Add Motion Sensor Light support (`gyd`). (#65)
- Add Multiple Dimmer and Dimmer Switch support (`tgq`, `tgkg`). (#82)
- Add Humidifier support (`jsq`). (#89) Thanks @akaminsky-net for the contribution


### Added
- Add config validation during plugin initialization.
- Add instruction message for handling API errors.
- Add debounce in `BaseAccessory.sendCommands()` for better API request peformance.
- Persist `TuyaDeviceList.{uid}.json` for debugging. (#41)
- Add `homeWhitelist` option for whitelisting homes. (#84) Thanks @JulianLepinski for the contribution


### Fixed
- Fix 1004 signature error when url query has more than 2 elements.
- Fix 1010 token expired error when refresh access_token.
- Fix 1106 permission error when polling device info list.
- Fix 1100, 2017 errors when login. (via config validation)
- Fix Lightbulb `RGBW` and `RGBCW` work mode not switched properly (#12 #56 #59)
- Fix Lightbulb color temperature not working. (#13)
- Fix Thermostat temperature units handling. (#20)
- Fix Thermostat mode handling. (#26)
- Fix Curtain Switch with no position feature. (#27)
- Fallback when receiving MQTT message with wrong order. (#35)
- Fix wrong temperature on sensor. (#38)
- Fix fan speed issue. (#46 #51)
- Workaround for Thermostat with wrong schema property (#74)
- Fix Contact Sensor not working (#75)
- Fix iOS 16 default accessory name issue. (#85)


### Changed
- Rewritten in TypeScript, brings benefits of type checking, smart code hints, etc.
- Reimplement accessory logics. More friendly for accessory developers.
- Update device info list polling logic. Less API errors.
- Now `Manufactor`, `Serial Number` and `Model` will be correctly displayed in HomeKit.
- All devices will be shown in HomeKit by default (Including unsupported device).
- Updated unit test.
- Updated documentations. Thanks @prabch for the contribution


### Removed
- Remove `debug` option. Silence logs for users. For debugging, please refer to [troubleshooting](https://github.com/0x5e/homebridge-tuya-platform#troubleshooting).
- Remove `lang` option.
- Remove `username` and `password` options for `Custom` project. User will be created and authorized automatically. (#11)


src/config.ts
import { PlatformConfig } from 'homebridge';

export interface TuyaPlatformCustomConfigOptions {
  projectType: '1';
  endpoint: string;
  accessId: string;
  accessKey: string;
}

export interface TuyaPlatformHomeConfigOptions {
  projectType: '2';
  endpoint: string;
  accessId: string;
  accessKey: string;
}

export type TuyaPlatformConfigOptions = TuyaPlatformCustomConfigOptions | TuyaPlatformHomeConfigOptions;

export interface TuyaPlatformConfig extends PlatformConfig {
  options: TuyaPlatformConfigOptions;
}

export const customOptionsSchema = {
  properties: {
    endpoint: { type: 'string', format: 'url', required: true },
    accessId: { type: 'string', required: true },
    accessKey: { type: 'string', required: true },
  },
};

export const homeOptionsSchema = {
  properties: {
    endpoint: { type: 'string', format: 'url', required: true },
    accessId: { type: 'string', required: true },
    accessKey: { type: 'string', required: true },
  },
};


src/platform.ts
import { API, DynamicPlatformPlugin, Logger, PlatformAccessory, PlatformConfig, Service, Characteristic } from 'homebridge';
import { Validator } from 'jsonschema';

import TuyaOpenMQ from './core/TuyaOpenMQ';
import TuyaDevice, { TuyaDeviceStatus } from './device/TuyaDevice';
import TuyaDeviceManager from './device/TuyaDeviceManager';
import TuyaCustomDeviceManager from './device/TuyaCustomDeviceManager';
import TuyaHomeDeviceManager from './device/TuyaHomeDeviceManager';

import { PLATFORM_NAME, PLUGIN_NAME } from './settings';
import { TuyaPlatformConfigOptions, customOptionsSchema, homeOptionsSchema } from './config';
import AccessoryFactory from './accessory/AccessoryFactory';
import BaseAccessory from './accessory/BaseAccessory';
import TuyaOpenAPI, { LOGIN_ERROR_MESSAGES } from './core/TuyaOpenAPI';


/**
 * HomebridgePlatform
 * This class is the main constructor for your plugin, this is where you should
 * parse the user config and discover/register accessories with Homebridge.
 */
export class TuyaPlatform implements DynamicPlatformPlugin {
  public readonly Service: typeof Service = this.api.hap.Service;
  public readonly Characteristic: typeof Characteristic = this.api.hap.Characteristic;

  public options = this.config.options as TuyaPlatformConfigOptions;

  // this is used to track restored cached accessories
  public cachedAccessories: PlatformAccessory[] = [];

  public deviceManager?: TuyaDeviceManager;
  public accessoryHandlers: BaseAccessory[] = [];

  validate(config) {
    let result;
    if (!config.options) {
      this.log.warn('Not configured, exit.');
      return false;
    } else if (config.options.projectType === '1') {
      result = new Validator().validate(config.options, customOptionsSchema);
    } else if (config.options.projectType === '2') {
      result = new Validator().validate(config.options, homeOptionsSchema);
    } else {
      this.log.warn(`Unsupported projectType: ${config.options.projectType}, exit.`);
      return false;
    }
    result.errors.forEach(error => this.log.error(error.stack));
    return result.errors.length === 0;
  }

  constructor(
    public readonly log: Logger,
    public readonly config: PlatformConfig,
    public readonly api: API,
  ) {

    if (!this.validate(config)) {
      return;
    }

    this.log.debug('Finished initializing platform');

    // When this event is fired it means Homebridge has restored all cached accessories from disk.
    // Dynamic Platform plugins should only register new accessories after this event was fired,
    // in order to ensure they weren't added to homebridge already. This event can also be used
    // to start discovery of new accessories.
    this.api.on('didFinishLaunching', async () => {
      log.debug('Executed didFinishLaunching callback');
      // run the method to discover / register your devices as accessories
      await this.initDevices();
    });
  }

  /**
   * This function is invoked when homebridge restores cached accessories from disk at startup.
   * It should be used to setup event handlers for characteristics and update respective values.
   */
  configureAccessory(accessory: PlatformAccessory) {
    this.log.info('Loading accessory from cache:', accessory.displayName);

    // add the restored accessory to the accessories cache so we can track if it has already been registered
    this.cachedAccessories.push(accessory);
  }

  /**
   * This is an example method showing how to register discovered accessories.
   * Accessories must only be registered once, previously created accessories
   * must not be registered again to prevent "duplicate UUID" errors.
   */
  async initDevices() {

    let devices;
    if (this.options.projectType === '1') {
      devices = await this.initCustomProject();
    } else if (this.options.projectType === '2') {
      devices = await this.initHomeProject();
    } else {
      this.log.warn(`Unsupported projectType: ${this.config.options.projectType}.`);
    }

    if (!devices) {
      return;
    }

    // add accessories
    this.log.info(`Got ${devices.length} device(s).`);
    for (const device of devices) {
      this.addAccessory(device);
    }

    // remove unused accessories
    for (const cachedAccessory of this.cachedAccessories) {
      this.log.warn('Removing unused accessory from cache:', cachedAccessory.displayName);
      this.api.unregisterPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, [cachedAccessory]);
    }
    this.cachedAccessories = [];

    this.deviceManager!.on(TuyaDeviceManager.Events.DEVICE_ADD, this.addAccessory.bind(this));
    this.deviceManager!.on(TuyaDeviceManager.Events.DEVICE_INFO_UPDATE, this.updateAccessoryInfo.bind(this));
    this.deviceManager!.on(TuyaDeviceManager.Events.DEVICE_STATUS_UPDATE, this.updateAccessoryStatus.bind(this));
    this.deviceManager!.on(TuyaDeviceManager.Events.DEVICE_DELETE, this.removeAccessory.bind(this));

  }

  async initCustomProject() {
    if (this.options.projectType !== '1') {
      return null;
    }

    const DEFAULT_USER = 'homebridge';
    const DEFAULT_PASS = 'homebridge';

    let res;
    const { endpoint, accessId, accessKey } = this.options;
    const api = new TuyaOpenAPI(endpoint, accessId, accessKey, this.log);
    const mq = new TuyaOpenMQ(api, '2.0', this.log);
    const deviceManager = new TuyaCustomDeviceManager(api, mq);

    this.log.info('Get token.');
    res = await api.getToken();
    if (res.success === false) {
      this.log.error(`Get token failed. code=${res.code}, msg=${res.msg}`);
      return null;
    }


    this.log.info(`Search default user "${DEFAULT_USER}"`);
    res = await api.customGetUserInfo(DEFAULT_USER);
    if (res.success === false) {
      this.log.error(`Search user failed. code=${res.code}, msg=${res.msg}`);
      return null;
    }


    if (!res.result.user_name) {
      this.log.info(`Default user "${DEFAULT_USER}" not exist.`);
      this.log.info(`Creating default user "${DEFAULT_USER}".`);
      res = await api.customCreateUser(DEFAULT_USER, DEFAULT_PASS);
      if (res.success === false) {
        this.log.error(`Create default user failed. code=${res.code}, msg=${res.msg}`);
        return null;
      }
    } else {
      this.log.info(`Default user "${DEFAULT_USER}" exists.`);
    }
    const uid = res.result.user_id;


    this.log.info('Fetching asset list.');
    res = await deviceManager.getAssetList();
    if (res.success === false) {
      this.log.error(`Fetching asset list failed. code=${res.code}, msg=${res.msg}`);
      return null;
    }

    const assetIDList: string[] = [];
    for (const { asset_id, asset_name } of res.result.list) {
      this.log.info(`Got asset_id=${asset_id}, asset_name=${asset_name}`);
      assetIDList.push(asset_id);
    }

    if (assetIDList.length === 0) {
      this.log.warn('Asset list is empty. exit.');
      return null;
    }


    this.log.info('Authorize asset list.');
    res = await deviceManager.authorizeAssetList(uid, assetIDList, true);
    if (res.success === false) {
      this.log.error(`Authorize asset list failed. code=${res.code}, msg=${res.msg}`);
      return null;
    }


    this.log.info(`Log in with user "${DEFAULT_USER}".`);
    res = await api.customLogin(DEFAULT_USER, DEFAULT_USER);
    if (res.success === false) {
      this.log.error(`Login failed. code=${res.code}, msg=${res.msg}`);
      if (LOGIN_ERROR_MESSAGES[res.code]) {
        this.log.error(LOGIN_ERROR_MESSAGES[res.code]);
      }
      return null;
    }

    this.log.info('Start MQTT connection.');
    mq.start();

    this.log.info('Fetching device list.');
    const devices = await deviceManager.updateDevices(assetIDList);

    this.deviceManager = deviceManager;
    return devices;
  }

  async initHomeProject() {
    if (this.options.projectType !== '2') {
      return null;
    }

    let res;
    const { endpoint, accessId, accessKey } = this.options;
    const api = new TuyaOpenAPI(endpoint, accessId, accessKey, this.log);
    const mq = new TuyaOpenMQ(api, '1.0', this.log);
    const deviceManager = new TuyaHomeDeviceManager(api, mq);

    this.log.info('Get token.');
    res = await api.getToken();
    if (res.success === false) {
      this.log.error(`Get token failed. code=${res.code}, msg=${res.msg}`);
      return null;
    }
    const uid = api.tokenInfo.uid;

    this.log.info('Get user info.');
    res = await api.homeGetUserInfo(uid);
    if (res.success === false) {
      this.log.error(`Get user info failed. code=${res.code}, msg=${res.msg}`);
      return null;
    }

    // this.log.info('Log in to Tuya Cloud.');
    // res = await api.homeLogin(countryCode, username, password, appSchema);
    // if (res.success === false) {
    //   this.log.error(`Login failed. code=${res.code}, msg=${res.msg}`);
    //   if (LOGIN_ERROR_MESSAGES[res.code]) {
    //     this.log.error(LOGIN_ERROR_MESSAGES[res.code]);
    //   }
    //   return null;
    // }

    this.log.info('Start MQTT connection.');
    mq.start();

    this.log.info('Fetching home list.');
    res = await deviceManager.getHomeList(uid);
    if (res.success === false) {
      this.log.error(`Fetching home list failed. code=${res.code}, msg=${res.msg}`);
      return null;
    }

    const homeIDList: number[] = [];
    for (const { home_id, name } of res.result) {
      this.log.info(`Got home_id=${home_id}, name=${name}`);
      homeIDList.push(home_id);
    }

    if (homeIDList.length === 0) {
      this.log.warn('Home list is empty. exit.');
      return null;
    }

    this.log.info('Fetching device list.');
    const devices = await deviceManager.updateDevices(homeIDList);

    this.deviceManager = deviceManager;
    return devices;
  }

  addAccessory(device: TuyaDevice) {

    const uuid = this.api.hap.uuid.generate(device.id);
    const existingAccessory = this.cachedAccessories.find(accessory => accessory.UUID === uuid);
    if (existingAccessory) {
      this.log.info('Restoring existing accessory from cache:', existingAccessory.displayName);

      // Update context
      if (!existingAccessory.context || !existingAccessory.context.deviceID) {
        this.log.info('Update accessory context:', existingAccessory.displayName);
        existingAccessory.context.deviceID = device.id;
        this.api.updatePlatformAccessories([existingAccessory]);
      }

      // create the accessory handler for the restored accessory
      const handler = AccessoryFactory.createAccessory(this, existingAccessory, device);
      this.accessoryHandlers.push(handler);

      const index = this.cachedAccessories.indexOf(existingAccessory);
      if (index >= 0) {
        this.cachedAccessories.splice(index, 1);
      }

    } else {
      // the accessory does not yet exist, so we need to create it
      this.log.info('Adding new accessory:', device.name);

      // create a new accessory
      const accessory = new this.api.platformAccessory(device.name, uuid);
      accessory.context.deviceID = device.id;

      // create the accessory handler for the newly create accessory
      const handler = AccessoryFactory.createAccessory(this, accessory, device);
      this.accessoryHandlers.push(handler);

      // link the accessory to your platform
      this.api.registerPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, [accessory]);
    }
  }

  updateAccessoryInfo(device: TuyaDevice, info) {
    const handler = this.getAccessoryHandler(device.id);
    if (!handler) {
      return;
    }

    this.log.debug(`onDeviceInfoUpdate devId=${device.id}, info=${JSON.stringify(info)}`);
    handler.onDeviceInfoUpdate(info);
  }

  updateAccessoryStatus(device: TuyaDevice, status: TuyaDeviceStatus[]) {
    const handler = this.getAccessoryHandler(device.id);
    if (!handler) {
      return;
    }

    this.log.debug(`onDeviceStatusUpdate devId=${device.id}, status=${JSON.stringify(status)}`);
    handler.onDeviceStatusUpdate(status);
  }

  removeAccessory(deviceID: string) {
    const handler = this.getAccessoryHandler(deviceID);
    if (!handler) {
      return;
    }

    const index = this.accessoryHandlers.indexOf(handler);
    if (index >= 0) {
      this.accessoryHandlers.splice(index, 1);
    }

    this.api.unregisterPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, [handler.accessory]);
    this.log.info('Removing existing accessory from cache:', handler.accessory.displayName);
  }

  getAccessoryHandler(deviceID: string) {
    return this.accessoryHandlers.find(handler => handler.device.id === deviceID);
  }

}


src/device/TuyaHomeDeviceManager.ts
import TuyaDevice, { TuyaDeviceFunction } from './TuyaDevice';
import TuyaDeviceManager from './TuyaDeviceManager';

export default class TuyaHomeDeviceManager extends TuyaDeviceManager {

  async getHomeList(uid = this.api.tokenInfo.uid) {
    const res = await this.api.get(`/v1.0/users/${uid}/homes`);
    return res;
  }

  async getHomeDeviceList(homeID: number) {
    const res = await this.api.get(`/v1.0/homes/${homeID}/devices`);
    return res;
  }

  async updateDevices(homeIDList: number[]) {

    let devices: TuyaDevice[] = [];
    for (const homeID of homeIDList) {
      const res = await this.getHomeDeviceList(homeID);
      devices = devices.concat((res.result as []).map(obj => new TuyaDevice(obj)));
    }
    if (devices.length === 0) {
      return [];
    }

    const devIds: string[] = [];
    for (const device of devices) {
      devIds.push(device.id);
    }

    const functions = await this.getDeviceListFunctions(devIds);

    for (const device of devices) {
      for (const item of functions) {
        if (device.product_id === item['product_id']) {
          device.functions = item['functions'] as TuyaDeviceFunction[];
          break;
        }
      }
      device.functions = device.functions || [];
    }

    this.devices = devices;
    return devices;
  }

}


src/core/TuyaOpenAPI.ts
/* eslint-disable max-len */
/* eslint-disable @typescript-eslint/no-empty-function */
/* eslint-disable @typescript-eslint/no-unused-vars */
import axios, { Method } from 'axios';
import Crypto from 'crypto-js';
import { v4 as uuidv4 } from 'uuid';

// eslint-disable-next-line
// @ts-ignore
import { version } from '../../package.json';

import Logger from '../util/Logger';

enum Endpoints {
  AMERICA = 'https://openapi.tuyaus.com',
  AMERICA_EAST = 'https://openapi-ueaz.tuyaus.com',
  CHINA = 'https://openapi.tuyacn.com',
  EUROPE = 'https://openapi.tuyaeu.com',
  EUROPE_WEST = 'https://openapi-weaz.tuyaeu.com',
  INDIA = 'https://openapi.tuyain.com',
}

const DEFAULT_ENDPOINTS = {
  [Endpoints.AMERICA.toString()]: [1, 51, 52, 54, 55, 56, 57, 58, 60, 62, 63, 64, 66, 81, 82, 84, 95, 239, 245, 246, 500, 502, 591, 593, 594, 595, 597, 598, 670, 672, 674, 675, 677, 678, 682, 683, 686, 690, 852, 853, 886, 970, 1721, 1787, 1809, 1829, 1849, 4779, 5999, 35818],
  [Endpoints.CHINA.toString()]: [86],
  [Endpoints.EUROPE.toString()]: [7, 20, 27, 30, 31, 32, 33, 34, 36, 39, 40, 41, 43, 44, 45, 46, 47, 48, 49, 61, 65, 90, 92, 93, 94, 212, 213, 216, 218, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 240, 241, 242, 243, 244, 248, 250, 251, 252, 253, 254, 255, 256, 257, 258, 260, 261, 262, 263, 264, 265, 266, 267, 268, 269, 291, 297, 298, 299, 350, 351, 352, 353, 354, 355, 356, 357, 358, 359, 370, 371, 372, 373, 374, 375, 376, 377, 378, 379, 380, 381, 382, 385, 386, 387, 389, 420, 421, 423, 501, 503, 504, 505, 506, 507, 508, 509, 590, 592, 596, 673, 676, 679, 680, 681, 685, 687, 688, 689, 691, 692, 855, 856, 880, 960, 961, 962, 964, 965, 966, 967, 968, 971, 972, 973, 974, 975, 976, 977, 992, 993, 994, 995, 996, 998, 1242, 1246, 1264, 1268, 1284, 1340, 1345, 1441, 1473, 1649, 1664, 1670, 1671, 1684, 1758, 1767, 1784, 1868, 1869, 1876],
  [Endpoints.INDIA.toString()]: [91],
};

export const LOGIN_ERROR_MESSAGES = {
  1004: 'Please make sure your endpoint, accessId, accessKey is right.',
  1104: 'Please make sure your endpoint, accessId, accessKey is right.',
  1106: 'Please make sure your countryCode, username, password, appSchema is correct, and app account is linked with cloud project.',
  2401: 'Username or password is wrong.',
  2406: 'Please make sure your cloud project is created after May 25, 2021.',
};

export const API_ERROR_MESSAGES = {
  28841002: 'API subscription expired. Please renew the API subscription at Tuya IoT Platform.',
  28841105: `
API not authorized. Please go to "Tuya IoT Platform -> Cloud -> Development -> Project -> Service API",
and Authorize the following APIs before using:
- Authorization Token Management
- Device Status Notification
- IoT Core
- Industry Project Client Service (for "Custom" project type)
`,
};

type TuyaOpenAPIResponseSuccess = {
  success: true;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  result: any;
  t: number;
  tid: string;
};

type TuyaOpenAPIResponseError = {
  success: false;
  result: unknown;
  code: number;
  msg: string;
  t: number;
  tid: string;
};

export type TuyaOpenAPIResponse = TuyaOpenAPIResponseSuccess | TuyaOpenAPIResponseError;

export default class TuyaOpenAPI {

  static readonly Endpoints = Endpoints;

  public assetIDArr: Array<string> = [];
  public deviceArr: Array<object> = [];

  public tokenInfo = { access_token: '', refresh_token: '', uid: '', expire: 0 };

  constructor(
    public endpoint: Endpoints | string,
    public accessId: string,
    public accessKey: string,
    public log: Logger = console,
    public lang = 'en',
  ) {

  }

  isLogin() {
    return this.tokenInfo.access_token.length > 0;
  }

  isTokenExpired() {
    return (this.tokenInfo.expire - 60 * 1000 <= new Date().getTime());
  }

  isTokenManagementAPI(path: string) {
    if (path.startsWith('/v1.0/token')) {
      return true;
    }
    return false;
  }

  async _refreshAccessTokenIfNeed(path: string) {
    if (!this.isLogin()) {
      return;
    }

    if (!this.isTokenExpired()) {
      return;
    }

    if (this.isTokenManagementAPI(path)) {
      return;
    }

    this.log.debug('Refresh access_token');
    const res = await this.get(`/v1.0/token/${this.tokenInfo.refresh_token}`);
    if (res.success === false) {
      this.log.error(`Refresh access_token failed. code=${res.code}, msg=${res.msg}`);
      return;
    }

    const { access_token, refresh_token, uid, expire_time } = res.result;
    this.tokenInfo = {
      access_token: access_token,
      refresh_token: refresh_token,
      uid: uid,
      expire: expire_time * 1000 + new Date().getTime(),
    };

  }

  /**
   * In 'Custom' project, get a token directly. (Login with admin)
   * Have permission on asset management, user management.
   * But lost some permission on device management.
   * @returns
   */
  async getToken() {
    const res = await this.get('/v1.0/token', { grant_type: 1 });
    if (res.success) {
      const { access_token, refresh_token, uid, expire_time } = res.result;
      this.tokenInfo = {
        access_token: access_token,
        refresh_token: refresh_token,
        uid: uid,
        expire: expire_time * 1000 + new Date().getTime(),
      };
    }
    return res;
  }

  /**
   * In 'Smart Home' project, Get user info by uid.
   * @param uid User ID
   * @returns
   */
  async homeGetUserInfo(uid: string) {
    const res = await this.get(`/v1.0/users/${uid}/infos`);
    return res;
  }

  /**
   * In 'Smart Home' project, login with App's user.
   * @param countryCode 2-digit Country Code
   * @param username Username
   * @param password Password
   * @param appSchema App Schema: 'tuyaSmart', 'smartlife'
   * @returns
   */
  async homeLogin(countryCode: number, username: string, password: string, appSchema: string) {

    for (const _endpoint of Object.keys(DEFAULT_ENDPOINTS)) {
      const countryCodeList = DEFAULT_ENDPOINTS[_endpoint];
      if (countryCodeList.includes(countryCode)) {
        this.endpoint = <Endpoints>_endpoint;
      }
    }

    this.tokenInfo = { access_token: '', refresh_token: '', uid: '', expire: 0 };
    const res = await this.post('/v1.0/iot-01/associated-users/actions/authorized-login', {
      country_code: countryCode,
      username: username,
      password: Crypto.MD5(password).toString(),
      schema: appSchema,
    });

    if (res.success) {
      const { access_token, refresh_token, uid, expire_time, platform_url } = res.result;
      this.endpoint = platform_url || this.endpoint;
      this.tokenInfo = {
        access_token: access_token,
        refresh_token: refresh_token,
        uid: uid,
        expire: expire_time * 1000 + new Date().getTime(),
      };
    }

    return res;
  }

  /**
   * In 'Custom' project, Search user by username.
   * @param username Username
   * @returns
   */
  async customGetUserInfo(username: string) {
    const res = await this.get(`/v1.2/iot-02/users/${username}`);
    return res;
  }

  /**
   * In 'Custom' project, create a user.
   * @param username Username
   * @param password Password
   * @param country_code Country Code (Useless)
   * @returns
   */
  async customCreateUser(username: string, password: string, country_code = 1) {
    const res = await this.post('/v1.0/iot-02/users', {
      username,
      password: Crypto.SHA256(password).toString().toLowerCase(),
      country_code,
    });
    return res;
  }

  /**
   * In 'Custom' project, login with user.
   * @param username Username
   * @param password Password
   * @returns
   */
  async customLogin(username: string, password: string) {
    this.tokenInfo = { access_token: '', refresh_token: '', uid: '', expire: 0 };
    const res = await this.post('/v1.0/iot-03/users/login', {
      username: username,
      password: Crypto.SHA256(password).toString().toLowerCase(),
    });

    if (res.success) {
      const { access_token, refresh_token, uid, expire } = res.result;
      this.tokenInfo = {
        access_token: access_token,
        refresh_token: refresh_token,
        uid: uid,
        expire: expire * 1000 + new Date().getTime(),
      };
    }

    return res;
  }

  async request(method: Method, path: string, params?, body?) {
    await this._refreshAccessTokenIfNeed(path);

    const now = new Date().getTime();
    const nonce = uuidv4();
    const accessToken = this.tokenInfo.access_token || '';
    const stringToSign = this._getStringToSign(method, path, params, body);
    const headers = {
      't': `${now}`,
      'client_id': this.accessId,
      'nonce': nonce,
      'Signature-Headers': 'client_id',
      'sign': this._getSign(this.accessId, this.accessKey, this.isTokenManagementAPI(path) ? '' : this.tokenInfo.access_token, now, nonce, stringToSign),
      'sign_method': 'HMAC-SHA256',
      'access_token': accessToken,
      'lang': this.lang,
      'dev_lang': 'javascript',
      'dev_channel': 'homebridge',
      'devVersion': version,
    };
    // eslint-disable-next-line max-len
    this.log.debug(`TuyaOpenAPI request: method = ${method}, endpoint = ${this.endpoint}, path = ${path}, params = ${JSON.stringify(params)}, body = ${JSON.stringify(body)}, headers = ${JSON.stringify(headers)}`);

    const res = await axios({
      baseURL: this.endpoint,
      url: path,
      method: method,
      headers: headers,
      params: params,
      data: body,
    });

    this.log.debug(`TuyaOpenAPI response: path = ${path}, data = ${JSON.stringify(res.data)}`);
    if (res.data && API_ERROR_MESSAGES[res.data.code]) {
      this.log.error(API_ERROR_MESSAGES[res.data.code]);
    }

    return res.data as TuyaOpenAPIResponse;
  }

  async get(path: string, params?) {
    return this.request('get', path, params, null);
  }

  async post(path: string, params?) {
    return this.request('post', path, null, params);
  }

  async delete(path: string, params?) {
    return this.request('delete', path, params, null);
  }

  _getSign(accessId: string, accessKey: string, accessToken = '', timestamp = 0, nonce: string, stringToSign: string) {
    const message = [accessId, accessToken, timestamp, nonce, stringToSign].join('');
    const hash = Crypto.HmacSHA256(message, accessKey);
    const sign = hash.toString().toUpperCase();
    return sign;
  }

  _getStringToSign(method: Method, path: string, params, body) {
    const httpMethod = method.toUpperCase();
    const bodyStream = body ? JSON.stringify(body) : '';
    const contentSHA256 = Crypto.SHA256(bodyStream);
    const headers = `client_id:${this.accessId}\n`;
    const url = this._getSignUrl(path, params);
    const result = [httpMethod, contentSHA256, headers, url].join('\n');
    return result;
  }

  _getSignUrl(path: string, params) {
    if (!params) {
      return path;
    }

    const sortedKeys = Object.keys(params).sort();
    const kv: string[] = [];
    for (const key of sortedKeys) {
      if (params[key] !== null && params[key] !== undefined) {
        kv.push(`${key}=${params[key]}`);
      }
    }
    const url = `${path}?${kv.join('&')}`;

    return url;
  }

}
